packages/grid/x-data-grid/src/hooks/features/keyboardNavigation/useGridKeyboardNavigation.ts

```diff
@@ -107,6 +109,42 @@ export const useGridKeyboardNavigation = (
     [currentPageRows],
   );
 
+  const goToRightCell = React.useCallback(
```

Comment:
Since there's no difference between left/right navigation in cells and column headers, would it make more sense to have more generic functions `getLeftColumnIndex` and `getRightColumnIndex` that would return the proper index depending on `theme.direction`?
Here's a diff of the proposed changes to better illustrate the idea:
```diff
diff --git a/packages/grid/x-data-grid/src/hooks/features/keyboardNavigation/useGridKeyboardNavigation.ts b/packages/grid/x-data-grid/src/hooks/features/keyboardNavigation/useGridKeyboardNavigation.ts
index 6ad074a..3b059ee 100644
--- a/packages/grid/x-data-grid/src/hooks/features/keyboardNavigation/useGridKeyboardNavigation.ts
+++ b/packages/grid/x-data-grid/src/hooks/features/keyboardNavigation/useGridKeyboardNavigation.ts
@@ -29,6 +29,52 @@ function enrichPageRowsWithPinnedRows(
   return [...(pinnedRows.top || []), ...rows, ...(pinnedRows.bottom || [])];
 }
 
+const getLeftColumnIndex = ({
+  currentColIndex,
+  firstColIndex,
+  lastColIndex,
+  direction,
+}: {
+  currentColIndex: number;
+  firstColIndex: number;
+  lastColIndex: number;
+  direction: 'rtl' | 'ltr';
+}) => {
+  if (direction === 'ltr') {
+    if (currentColIndex > firstColIndex) {
+      return currentColIndex - 1;
+    }
+  } else if (direction === 'rtl') {
+    if (currentColIndex < lastColIndex) {
+      return currentColIndex + 1;
+    }
+  }
+  return null;
+};
+
+const getRightColumnIndex = ({
+  currentColIndex,
+  firstColIndex,
+  lastColIndex,
+  direction,
+}: {
+  currentColIndex: number;
+  firstColIndex: number;
+  lastColIndex: number;
+  direction: 'rtl' | 'ltr';
+}) => {
+  if (direction === 'ltr') {
+    if (currentColIndex < lastColIndex) {
+      return currentColIndex + 1;
+    }
+  } else if (direction === 'rtl') {
+    if (currentColIndex > firstColIndex) {
+      return currentColIndex - 1;
+    }
+  }
+  return null;
+};
+
 /**
  * @requires useGridSorting (method) - can be after
  * @requires useGridFilter (state) - can be after
@@ -109,42 +155,6 @@ export const useGridKeyboardNavigation = (
     [currentPageRows],
   );
 
-  const goToRightCell = React.useCallback(
-    (colIndexBefore: number, lastColIndex: number, rowIndexBefore: number) => {
-      if (colIndexBefore < lastColIndex) {
-        goToCell(colIndexBefore + 1, getRowIdFromIndex(rowIndexBefore), 'right');
-      }
-    },
-    [goToCell, getRowIdFromIndex],
-  );
-
-  const goToLeftCell = React.useCallback(
-    (colIndexBefore: number, firstColIndex: number, rowIndexBefore: number) => {
-      if (colIndexBefore > firstColIndex) {
-        goToCell(colIndexBefore - 1, getRowIdFromIndex(rowIndexBefore));
-      }
-    },
-    [goToCell, getRowIdFromIndex],
-  );
-
-  const goToRightColumnHeader = React.useCallback(
-    (colIndexBefore: number, lastColIndex: number, event: React.KeyboardEvent<HTMLElement>) => {
-      if (colIndexBefore < lastColIndex) {
-        goToHeader(colIndexBefore + 1, event);
-      }
-    },
-    [goToHeader],
-  );
-
-  const goToLeftColumnHeader = React.useCallback(
-    (colIndexBefore: number, firstColIndex: number, event: React.KeyboardEvent<HTMLElement>) => {
-      if (colIndexBefore > firstColIndex) {
-        goToHeader(colIndexBefore - 1, event);
-      }
-    },
-    [goToHeader],
-  );
-
   const handleColumnHeaderKeyDown = React.useCallback<GridEventListener<'columnHeaderKeyDown'>>(
     (params, event) => {
       const headerTitleNode = event.currentTarget.querySelector(
@@ -182,19 +192,27 @@ export const useGridKeyboardNavigation = (
         }
 
         case 'ArrowRight': {
-          if (theme.direction === 'ltr') {
-            goToRightColumnHeader(colIndexBefore, lastColIndex, event);
-          } else if (theme.direction === 'rtl') {
-            goToLeftColumnHeader(colIndexBefore, firstColIndex, event);
+          const rightColIndex = getRightColumnIndex({
+            currentColIndex: colIndexBefore,
+            firstColIndex,
+            lastColIndex,
+            direction: theme.direction,
+          });
+          if (rightColIndex !== null) {
+            goToHeader(rightColIndex, event);
           }
           break;
         }
 
         case 'ArrowLeft': {
-          if (theme.direction === 'ltr') {
-            goToLeftColumnHeader(colIndexBefore, firstColIndex, event);
-          } else if (theme.direction === 'rtl') {
-            goToRightColumnHeader(colIndexBefore, lastColIndex, event);
+          const leftColIndex = getLeftColumnIndex({
+            currentColIndex: colIndexBefore,
+            firstColIndex,
+            lastColIndex,
+            direction: theme.direction,
+          });
+          if (leftColIndex !== null) {
+            goToHeader(leftColIndex, event);
           }
           break;
         }
@@ -252,13 +270,11 @@ export const useGridKeyboardNavigation = (
     [
       apiRef,
       currentPageRows.length,
-      theme.direction,
       goToCell,
       getRowIdFromIndex,
       goToHeader,
       goToGroupHeader,
-      goToRightColumnHeader,
-      goToLeftColumnHeader,
+      theme.direction,
     ],
   );
 
@@ -429,19 +445,35 @@ export const useGridKeyboardNavigation = (
         }
 
         case 'ArrowRight': {
-          if (direction === 'ltr') {
-            goToRightCell(colIndexBefore, lastColIndex, rowIndexBefore);
-          } else if (direction === 'rtl') {
-            goToLeftCell(colIndexBefore, firstColIndex, rowIndexBefore);
+          const rightColIndex = getRightColumnIndex({
+            currentColIndex: colIndexBefore,
+            firstColIndex,
+            lastColIndex,
+            direction,
+          });
+          if (rightColIndex !== null) {
+            goToCell(
+              rightColIndex,
+              getRowIdFromIndex(rowIndexBefore),
+              direction === 'rtl' ? 'left' : 'right',
+            );
           }
           break;
         }
 
         case 'ArrowLeft': {
-          if (direction === 'ltr') {
-            goToLeftCell(colIndexBefore, firstColIndex, rowIndexBefore);
-          } else if (direction === 'rtl') {
-            goToRightCell(colIndexBefore, lastColIndex, rowIndexBefore);
+          const leftColIndex = getLeftColumnIndex({
+            currentColIndex: colIndexBefore,
+            firstColIndex,
+            lastColIndex,
+            direction,
+          });
+          if (leftColIndex !== null) {
+            goToCell(
+              leftColIndex,
+              getRowIdFromIndex(rowIndexBefore),
+              direction === 'rtl' ? 'right' : 'left',
+            );
           }
           break;
         }
@@ -522,16 +554,7 @@ export const useGridKeyboardNavigation = (
         event.preventDefault();
       }
     },
-    [
-      apiRef,
-      currentPageRows,
-      theme.direction,
-      getRowIdFromIndex,
-      goToCell,
-      goToHeader,
-      goToRightCell,
-      goToLeftCell,
-    ],
+    [apiRef, currentPageRows, theme.direction, getRowIdFromIndex, goToCell, goToHeader],
   );
 
   useGridApiEventHandler(apiRef, 'columnHeaderKeyDown', handleColumnHeaderKeyDown);
```

What do you think?